## *Importing required libraries*

In [34]:
!pip install gensim
import pandas as pd
import gensim

## *Merged 2 excel sheets datasets, in total we have used 329 asanas benefits*

In [35]:
df = pd.read_csv('test.csv')

In [36]:
df

Asana  \
0                PADOTTHANASANA   
1                   PARVATASANA   
2            ARDHA TITALI ASANA   
3    GATYATMAK MERU \nVAKRASANA   
4              SIDEWAYS VIEWING   
..                          ...   
321                  kapilasana   
322                  omkarasana   
323               kashyapawsana   
324               bhunamanasana   
325                 Mandalasana   

                                              Benefits  
0    This asana strengthens\nthe abdominal muscles ...  
1    This pose strengthens the nerves and muscles i...  
2    This is an excellent \npreparatory practice fo...  
3    This asana removes stiffness \nof the back and...  
4    Sideways viewing relaxes the \ntension of the ...  
..                                                 ...  
321  1. Helps in stretching legs, hamstrings, arms,...  
322  It relieves the pain of hands and feet( It is ...  
323  1. This asana provides a deep muscular massage...  
324  1. Bhunamanasana stretches and improves the fl...  
325  1) Stretches, Strengthens and Lengthens: Manda...  

[326 rows x 2 columns]

In [54]:
poses = df["Asana"]
l = [] 
for i in poses:
    l.append(i)
print(l)

['PADOTTHANASANA', 'PARVATASANA', 'ARDHA TITALI ASANA', 'GATYATMAK MERU \nVAKRASANA', 'SIDEWAYS VIEWING', 'MAKARASANA', 'PADMASANA', 'VAJRASANA', 'ARDHA \nCHANDRASANA', 'YOGAMUDRASANA', 'BHUJANGASANA', 'SAITHALYASANA', 'BHU NAMANASANA', 'SARVANGASANA', 'NATARAJASANA', 'POORNA \nBHUJANGASANA', 'KOORMASANA', 'POORNA \nSHALABHASANA', 'POORNA \nDHANURASANA', 'BANDHA HASTA UTTHANASANA ', 'SHAVA UDARAKARSHANASANA ', 'CHAKKI CHALANASANA ', 'KASHTHA TAKSHANASANA ', 'VAYU NISHKASANA', 'USHTRASANA', 'Samakonasana ', 'Matsyasana', 'Kandharasana', ' Setu Asana ', 'Paschimottanasana', 'Meru Akarshanasana', 'Pada Hastasana', 'Seetkari Pranayama', 'Jalandhara Bandha', 'Tadagi Mudra', 'Maha Vedha Mudra', 'Shashankasana', 'JANU CHAKRA', 'POORNA TITALI ASANA', 'MANIBANDHA CHAKRA', 'SKANDHA CHAKRA', 'GREEVA SANCHALANA', 'PADACHAKRASANA', 'PADA SANCHALANASANA', 'SUPTA PAWANMUKTASANA', 'JHULANA LURHAKANASANA', 'SUPTA UDARAKARSHANASANA', 'NAUKASANA', 'RAJJU KARSHANASANA', 'GATYATMAK MERU VAKRASANA', 'NAUKA 

In [52]:
pip install pygoogle_image

In [53]:
import pygoogle_image
from pygoogle_image import image as pi
for i in poses:
    pi.download(i,limit=1)

[========================================================================] 100%
[========================================================================] 100%
[========================================================================] 100%


OSError: [WinError 123] The filename, directory name, or volume label syntax is incorrect: 'images/GATYATMAK_MERU_\nVAKRASANA'

In [4]:
#shape of the data frame
df.shape

(326, 2)

## *Removing Stop words at first and converting it into lower cases so that all stop words can be remove which were in Upper cases.*


In [5]:
! pip install nltk
import nltk
from nltk.corpus import stopwords
nltk.download('stopwords')
stop = stopwords.words('english')
#  1st step for removing stop words is to use the library and remove stop words, this will make 
# sure that some basic stop words and numerical values is been removed from the Benefits column.

# 2nd step is that there might be stop words present in upper case for example "This" is a stop words 
# which was not removed in 1st step so I converted the Benefits column to lower case. Moreover we can only  
# lowercase any sentence or entire column when that column is free of any numerical value.

# 3rd then further removing stopwords, in this way we get the whole benefits column free from stop words


from gensim.parsing.preprocessing import remove_stopwords
from gensim.parsing.preprocessing import strip_non_alphanum
from gensim.parsing.preprocessing import strip_numeric
from gensim.parsing.preprocessing import strip_multiple_whitespaces
df['Benefits'] = df['Benefits'].apply(str.lower)
df['Benefits']=df.Benefits.apply(remove_stopwords)
df['Benefits']=df.Benefits.apply(strip_non_alphanum)
df['Benefits']=df.Benefits.apply(strip_numeric)
df['Benefits']=df.Benefits.apply(strip_multiple_whitespaces)
df['Asana']=df.Asana.apply(strip_multiple_whitespaces)
df['Asana'] = df['Asana'].apply(str.lower)
df

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\91950\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


Asana  \
0              padotthanasana   
1                 parvatasana   
2          ardha titali asana   
3    gatyatmak meru vakrasana   
4            sideways viewing   
..                        ...   
321                kapilasana   
322                omkarasana   
323             kashyapawsana   
324             bhunamanasana   
325               mandalasana   

                                              Benefits  
0    asana strengthens abdominal muscles massages o...  
1    pose strengthens nerves muscles limbs back hel...  
2    excellent preparatory practice loosening knee ...  
3    asana removes stiffness increases flexibility ...  
4    sideways viewing relaxes tension muscles strai...  
..                                                 ...  
321   helps stretching legs hamstrings arms chest b...  
322  relieves pain hands feet helpful rheumatism gi...  
323   asana provides deep muscular massage abdomen ...  
324   bhunamanasana stretches improves flexibility ...  
325   stretches strengthens lengthens mandalasana c...  

[326 rows x 2 columns]

## *We have further removed special characters and tokenized each row of benefits.* 

In [6]:
# this will create the list of each benefit rows
benefits =df['Benefits'].apply(gensim.utils.simple_preprocess)
print(benefits)

0      [asana, strengthens, abdominal, muscles, massa...
1      [pose, strengthens, nerves, muscles, limbs, ba...
2      [excellent, preparatory, practice, loosening, ...
3      [asana, removes, stiffness, increases, flexibi...
4      [sideways, viewing, relaxes, tension, muscles,...
                             ...                        
321    [helps, stretching, legs, hamstrings, arms, ch...
322    [relieves, pain, hands, feet, helpful, rheumat...
323    [asana, provides, deep, muscular, massage, abd...
324    [bhunamanasana, stretches, improves, flexibili...
325    [stretches, strengthens, lengthens, mandalasan...
Name: Benefits, Length: 326, dtype: object


In [7]:
# prints the words present in the benefit row 1
benefits[0]

['asana',
 'strengthens',
 'abdominal',
 'muscles',
 'massages',
 'organs',
 'strengthens',
 'digestive',
 'system',
 'lower',
 'back',
 'pelvic',
 'perineal',
 'muscles',
 'helps',
 'correct',
 'prolapse']

## *Implemented Word2Vec with following features:*
* window size = 5
* minimum word cound for which the model can input row wise data = 2
* required cpu  threads to train the model = 4
* size of the required vector embedding = 50

In [8]:

model = gensim.models.Word2Vec(
    window=5,
    min_count=2,
    workers=4
)


## *Building vocabulary of unique words present in the entire benefit column*

In [9]:

model.build_vocab(benefits, progress_per=5)
# vocab_len = len(model.wv)
# print(vocab_len)

In [10]:
model.train(benefits, total_examples=model.corpus_count, epochs=2000)

(22224124, 29294000)

## *We can test our model for any words suppose say sciatica we will get the similar words in benefits* 

In [11]:
#when we find the similar words for a disease in 
#Benefits section we also see that we get the names of asanas also.
#like for this word we get similar words like chakrasana, marjariasana, shalabhasana which indicates
#that since we have taken these words from benefits section, it means that for this particular diseaes
#most prpbable asanas can be chakrasana, marjariasana, shalabhasana etc.  to cure it.
model.wv.most_similar("sciatica", topn= 100)

#print(type(model.wv.most_similar("sciatica", topn= 100)))

[('waist', 0.3302907943725586),
 ('time', 0.31766268610954285),
 ('stiff', 0.3040863573551178),
 ('lumbago', 0.30091384053230286),
 ('disc', 0.3000882565975189),
 ('pains', 0.29202914237976074),
 ('spondylitis', 0.2905334532260895),
 ('sciatic', 0.2845100164413452),
 ('risk', 0.2750784456729889),
 ('treatment', 0.26699164509773254),
 ('slipped', 0.2626746594905853),
 ('rid', 0.26253700256347656),
 ('backache', 0.2590833008289337),
 ('mild', 0.24188394844532013),
 ('aches', 0.23763704299926758),
 ('hands', 0.23366311192512512),
 ('rounded', 0.23048242926597595),
 ('realign', 0.225270077586174),
 ('sacrum', 0.22315745055675507),
 ('vertebrae', 0.2113790363073349),
 ('marjariasana', 0.20928429067134857),
 ('lifts', 0.2068244218826294),
 ('prevents', 0.2044043242931366),
 ('discs', 0.2025301307439804),
 ('region', 0.2019566148519516),
 ('need', 0.20138733088970184),
 ('benefit', 0.20136398077011108),
 ('ovaries', 0.19954627752304077),
 ('certain', 0.19859151542186737),
 ('arms', 0.19331558

## *Printing one of the benefits column word's vector*

In [12]:
# suppose we want to know the vector embedding of a word 'pain' so we have printed the vector embedding of size 50
import numpy as np
print((model.wv.get_vector('pain')))

[  2.6980839    4.516741     2.943067     1.7197337   -0.3952227
   8.737232    -5.279915     1.3644663   -4.2818384   -0.52353895
   0.8957256   -0.675856     2.8455715   -2.3637404    1.1484559
  -0.7005269    2.1599581   -5.9692664   -1.5387658   -1.1188648
   3.0717564   -0.04725587  -0.8080487    1.8293533   -2.1144907
   2.2897837    7.6490965   -1.8065692   -2.873342    -1.310091
   4.844063     5.101049   -11.633857    -2.5063787    3.3282645
  -3.7794979    2.4778123    4.856119     0.7332692   -0.3851651
  -1.4146578   10.635022    -3.860674     0.7276645    1.1141173
   0.9794601   -5.435132    -3.8533902   -1.0369657    1.5819745
   4.4276667   -2.4250026    0.30966568  -0.32872435  -3.2640338
  -1.6665647   -5.2941117    8.319825     2.1889215   -7.1624274
   0.48382428  -4.1300344   -4.2772107    6.56779     -2.28764
  -0.80132586  -2.4751685    4.611089     2.7215319   -1.6889685
   2.8899324    6.7211676   -0.8095171    2.2351115    0.59228665
  -2.186682     2.6760113 

In [13]:
# counts total number of rows in datasets being trained
model.corpus_count

326

In [14]:
model.save("model.bin")

In [15]:
from gensim.models import Word2Vec

# Load the Word2Vec model
model = Word2Vec.load("model.bin")

# Get the list of unique words in the vocabulary
words = list(model.wv.key_to_index.keys())


In [16]:
# printing unique words from the vocabulary list.
print(words)

['muscles', 'body', 'pose', 'helps', 'organs', 'asana', 'spine', 'improves', 'abdominal', 'hips', 'shoulders', 'strengthens', 'blood', 'practice', 'lower', 'balance', 'stretches', 'stretch', 'tones', 'chest', 'legs', 'flexibility', 'neck', 'good', 'arms', 'posture', 'abdomen', 'nervous', 'mind', 'back', 'flow', 'yoga', 'circulation', 'reproductive', 'pelvic', 'system', 'leg', 'hip', 'entire', 'stimulates', 'awareness', 'nerves', 'spinal', 'core', 'improving', 'deep', 'digestive', 'heart', 'gives', 'digestion', 'joints', 'pressure', 'benefits', 'chakra', 'increases', 'strength', 'concentration', 'great', 'upper', 'energy', 'constipation', 'sense', 'thighs', 'functioning', 'like', 'related', 'toning', 'region', 'glands', 'stretching', 'knees', 'breathing', 'toned', 'especially', 'poses', 'focus', 'hamstrings', 'breath', 'strong', 'massages', 'area', 'liver', 'ankles', 'internal', 'better', 'help', 'improve', 'pain', 'the', 'alignment', 'brain', 'disorders', 'stability', 'flexible', 'leve

In [17]:
# here we have the length of unique words

print(len(words))

1330


## *Created an empty dictionary at first to store the unique words as key words  along with its vector embeddings.*

In [18]:
dict_of_word_embeddings = dict({})
for i in words:
    dict_of_word_embeddings[i] = model.wv[i]

In [19]:
# on printing the dictionary, we get the following result.
print(dict_of_word_embeddings)

{'muscles': array([-0.76654214,  2.900865  ,  0.5912934 ,  2.2766824 ,  0.1380029 ,
       -0.5027816 , -2.1136012 ,  0.81286967, -2.8088787 ,  1.8234394 ,
        0.18834877, -3.8951244 , -1.1333603 ,  1.7753909 ,  0.23096174,
       -4.0830083 , -1.2106866 ,  0.40549695, -2.0672386 ,  0.26444578,
        2.353652  ,  2.234708  ,  1.7484134 ,  2.4196894 , -0.13229862,
        0.63491434, -0.57346094,  0.6684417 , -1.1212571 ,  0.20886573,
        2.950679  ,  0.57606566, -0.30090737,  2.6207943 , -2.7648475 ,
        0.4391897 ,  0.8784108 , -1.3114974 , -1.1144582 , -3.7010498 ,
        0.21621047,  0.01498064,  1.1208699 , -1.3100672 ,  1.3213575 ,
        0.41055402, -0.47351766,  3.0583386 , -0.16944079,  1.8412448 ,
        0.17517154,  1.3839892 , -0.94374543,  1.0965972 ,  0.00474672,
        2.6201873 ,  2.6105034 , -0.8876475 , -1.1957086 , -0.8030171 ,
        2.4645562 ,  0.6616804 ,  0.8484845 , -0.6754273 , -3.4193354 ,
       -0.98577   ,  0.5666034 , -0.8883875 , -0.286

## *Exporting dictionary data into separate excel sheet/ csv file*.

In [20]:
Unique_words = dict_of_word_embeddings.keys()
word_vectors  = dict_of_word_embeddings.values()
# print(asanas, word_vectors)
d = {'Unique_words' : Unique_words , 'Word_Vectors' : word_vectors}
dataframe = pd.DataFrame(data = d)
dataframe

Unique_words                                       Word_Vectors
0         muscles  [-0.76654214, 2.900865, 0.5912934, 2.2766824, ...
1            body  [-1.3609003, 0.28981966, 2.5230618, -0.1953817...
2            pose  [0.12473381, 1.2486181, 0.033645216, -2.497720...
3           helps  [-2.1865203, 0.8846974, 1.319828, 0.84958655, ...
4          organs  [-2.32748, 1.0284241, -3.1051574, -1.1932904, ...
...           ...                                                ...
1325       little  [0.9713384, 0.33615562, -0.8301823, 0.64237475...
1326  trikonasana  [2.6214528, 4.0123396, -2.7287693, 2.5857074, ...
1327     migraine  [1.8771211, -0.8430638, 0.97005874, 3.8555207,...
1328        heals  [2.7012422, 0.5879855, 1.1658524, -0.019796861...
1329        short  [3.0078633, -0.562787, 1.6616246, 3.2130675, -...

[1330 rows x 2 columns]

## *We have made list of unique asanas as after merging excel sheets there were repeated asanas*

In [21]:
asanas = list(df['Asana'])
# total asanas present ( with repetition)
print(len(asanas))
asana = []
      

for x in asanas:
  if x not in asana:
    asana.append(x)
# total number of unique asanas
print(len(asana))    
# list of unique asanas
print(asana)

326
293
['padotthanasana', 'parvatasana', 'ardha titali asana', 'gatyatmak meru vakrasana', 'sideways viewing', 'makarasana', 'padmasana', 'vajrasana', 'ardha chandrasana', 'yogamudrasana', 'bhujangasana', 'saithalyasana', 'bhu namanasana', 'sarvangasana', 'natarajasana', 'poorna bhujangasana', 'koormasana', 'poorna shalabhasana', 'poorna dhanurasana', 'bandha hasta utthanasana ', 'shava udarakarshanasana ', 'chakki chalanasana ', 'kashtha takshanasana ', 'vayu nishkasana', 'ushtrasana', 'samakonasana ', 'matsyasana', 'kandharasana', ' setu asana ', 'paschimottanasana', 'meru akarshanasana', 'pada hastasana', 'seetkari pranayama', 'jalandhara bandha', 'tadagi mudra', 'maha vedha mudra', 'shashankasana', 'janu chakra', 'poorna titali asana', 'manibandha chakra', 'skandha chakra', 'greeva sanchalana', 'padachakrasana', 'pada sanchalanasana', 'supta pawanmuktasana', 'jhulana lurhakanasana', 'supta udarakarshanasana', 'naukasana', 'rajju karshanasana', 'nauka sanchalanasana', 'namaskarasan

In [22]:
!pip install scikit-learn

## *On-hot Encodded the asanas words and created a dictionary to store the asana word with its word embedding*

In [23]:
from tensorflow import keras
from tensorflow.keras.utils import to_categorical
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder
label_encoder = LabelEncoder()
integer_encoded = label_encoder.fit_transform(asana)
#print(integer_encoded)
integer_encoded = integer_encoded.reshape(len(integer_encoded), 1)

### One hot encoding
onehot_encoder = OneHotEncoder(sparse=False)
onehot_encoded = onehot_encoder.fit_transform(integer_encoded)

# onehot_encoded

asan_dict={}

for i in range(len(asana)):
  asan_dict[asana[i]] = onehot_encoded[i]

print(asan_dict) 


C:\Users\91950\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\preprocessing\_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


{'padotthanasana': array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
      

In [24]:
from tensorflow import keras
from tensorflow.keras.layers import Dense

In [25]:
# printing the benefit words of one row.
print(benefits[0])

['asana', 'strengthens', 'abdominal', 'muscles', 'massages', 'organs', 'strengthens', 'digestive', 'system', 'lower', 'back', 'pelvic', 'perineal', 'muscles', 'helps', 'correct', 'prolapse']


## *Created a Tupple associating the benefits words of each row with the respective asana. In order to bring the associativity between benefits words and asana words*

In [26]:
pair=[]

i=0
a=len(asana)
for x in benefits:
  if(i<a):
    target=asana[i]
    for y in x:
      if(y not in words):
        continue
      pair.append((y,target))
  i+=1  
print(pair)




[('asana', 'padotthanasana'), ('strengthens', 'padotthanasana'), ('abdominal', 'padotthanasana'), ('muscles', 'padotthanasana'), ('massages', 'padotthanasana'), ('organs', 'padotthanasana'), ('strengthens', 'padotthanasana'), ('digestive', 'padotthanasana'), ('system', 'padotthanasana'), ('lower', 'padotthanasana'), ('back', 'padotthanasana'), ('pelvic', 'padotthanasana'), ('muscles', 'padotthanasana'), ('helps', 'padotthanasana'), ('correct', 'padotthanasana'), ('prolapse', 'padotthanasana'), ('pose', 'parvatasana'), ('strengthens', 'parvatasana'), ('nerves', 'parvatasana'), ('muscles', 'parvatasana'), ('limbs', 'parvatasana'), ('back', 'parvatasana'), ('helps', 'parvatasana'), ('increase', 'parvatasana'), ('height', 'parvatasana'), ('stretching', 'parvatasana'), ('muscles', 'parvatasana'), ('ligaments', 'parvatasana'), ('enabling', 'parvatasana'), ('growing', 'parvatasana'), ('bones', 'parvatasana'), ('grow', 'parvatasana'), ('longer', 'parvatasana'), ('circulation', 'parvatasana'), 

## *Making a 2-D array of context words(benefit words) and Target words(asana words) by numpy stack*

In [27]:
contexts=[dict_of_word_embeddings[context] for context,target in pair]
contexts=np.vstack(contexts)
# shape of the context words matrix
contexts.shape


(12085, 100)

In [28]:
targets=[asan_dict[target] for context,target in pair]
targets=np.vstack(targets)
# shape of the target words matrix

targets.shape

(12085, 293)

## *Implemented **Artificial Neural Network**. with the help of tensorfow's keras functional API with the following features:*

* Number of input layer = 1
* Size of input layer = 50 unit
* Number of dense layer = 1
* size of dense layer = 1000 units
* Number of output layer = 1
* size of output unit = 293
* activation function = softmax
* loss = categorical_crossentropy
* optimizer = adam
* number of epotchs = 100

In [29]:
from tensorflow import keras
from tensorflow.keras.layers import Dense

network_input = keras.Input(shape=contexts.shape[1], name='input_layer')
                                                                                    # Create a hidden layer for the network; store under 'hidden_layer'
hidden_layer1 = Dense(units=1000, activation='sigmoid', name='hidden_layer1')(network_input)

                                                                                            # Create an output layer for the network; store under 'output_layer'
output_layer = Dense(units=targets.shape[1], activation='softmax', name='output_layer')(hidden_layer1)

                                                                          # Create a Keras Model; store under 'embedding_model'
embedding_model = keras.Model(inputs=network_input, outputs=output_layer)

                                                          # Compile the model for training; define loss function
embedding_model.compile(loss='categorical_crossentropy',optimizer = 'adam', metrics = ['accuracy'])

                                                          # Print out a summary of the model
embedding_model.summary()

Model: "model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_layer (InputLayer)    [(None, 100)]             0         
                                                                 
 hidden_layer1 (Dense)       (None, 1000)              101000    
                                                                 
 output_layer (Dense)        (None, 293)               293293    
                                                                 
Total params: 394,293
Trainable params: 394,293
Non-trainable params: 0
_________________________________________________________________


In [30]:
# Fit a model to the data
embedding_model.fit(x=contexts,   # inputs
                    y=targets,   # outputs
                    batch_size=1024,  # how many pairs of words processed simultaneously
                    epochs=100,   # how many times we loop through the whole data
                    verbose=1   # do not print training status
                   )



Epoch 1/100
12/12 [==============================] - 1s 22ms/step - loss: 5.4632 - accuracy: 0.0187
Epoch 2/100
12/12 [==============================] - 0s 21ms/step - loss: 5.1180 - accuracy: 0.0460
Epoch 3/100
12/12 [==============================] - 0s 22ms/step - loss: 4.8949 - accuracy: 0.0671
Epoch 4/100
12/12 [==============================] - 0s 25ms/step - loss: 4.7155 - accuracy: 0.0879
Epoch 5/100
12/12 [==============================] - 0s 27ms/step - loss: 4.5656 - accuracy: 0.1010
Epoch 6/100
12/12 [==============================] - 0s 26ms/step - loss: 4.4418 - accuracy: 0.1087
Epoch 7/100
12/12 [==============================] - 0s 29ms/step - loss: 4.3371 - accuracy: 0.1144
Epoch 8/100
12/12 [==============================] - 0s 22ms/step - loss: 4.2447 - accuracy: 0.1206
Epoch 9/100
12/12 [==============================] - 0s 23ms/step - loss: 4.1656 - accuracy: 0.1247
Epoch 10/100
12/12 [==============================] - 0s 27ms/step - loss: 4.0940 - accuracy: 0.1289

12/12 [==============================] - 0s 26ms/step - loss: 3.0437 - accuracy: 0.1556
Epoch 83/100
12/12 [==============================] - 0s 29ms/step - loss: 3.0386 - accuracy: 0.1576
Epoch 84/100
12/12 [==============================] - 0s 34ms/step - loss: 3.0344 - accuracy: 0.1572
Epoch 85/100
12/12 [==============================] - 0s 36ms/step - loss: 3.0326 - accuracy: 0.1571
Epoch 86/100
12/12 [==============================] - 0s 33ms/step - loss: 3.0326 - accuracy: 0.1562
Epoch 87/100
12/12 [==============================] - 0s 31ms/step - loss: 3.0283 - accuracy: 0.1604
Epoch 88/100
12/12 [==============================] - 0s 29ms/step - loss: 3.0265 - accuracy: 0.1582
Epoch 89/100
12/12 [==============================] - 0s 26ms/step - loss: 3.0214 - accuracy: 0.1564
Epoch 90/100
12/12 [==============================] - 0s 29ms/step - loss: 3.0201 - accuracy: 0.1568
Epoch 91/100
12/12 [==============================] - 0s 28ms/step - loss: 3.0152 - accuracy: 0.1577
Epo

In [31]:
import pickle
filename = 'savemodel.sav'
pickle.dump(embedding_model,open(filename,'wb'))

In [32]:
load_model = pickle.load(open(filename,'rb'))

## *Function which inputs the user details and suggests user with the most recommended asanas*

In [33]:
from collections import Counter
from IPython.display import clear_output

def magic():
  predicted_asanas = []
  user_input_words= []
  final_predicted_asanas = []
  number_in_words = ['first','second', 'third', 'fourth']
  for i in range(4):
    user_input_words.append(input(f"Enter {number_in_words[i]} benefit word:  "))
  for i in user_input_words:
    if i in dict_of_word_embeddings:

      input_array = np.expand_dims(dict_of_word_embeddings[i], axis=0)
      prediction = load_model.predict(input_array)
      flatten_pred = prediction.flatten()
      result_indices = flatten_pred.argsort()[-10:][::-1]
    
      for result in result_indices:
        predicted_asanas.append(asana[result])
    
    
  counter_found = Counter(predicted_asanas)
  final_predicted_asanas_with_freq = counter_found.most_common(7)
  # print(final_predicted_asanas_with_freq)

  for yoga, freq in final_predicted_asanas_with_freq:
    final_predicted_asanas.append(yoga)
  
  print(final_predicted_asanas)
  choice=input("Clear output: Y/N ")
  if choice=='Y':
   clear_output()
  
  
magic()


KeyboardInterrupt

